File: leetcode/20.ts
//*
// Find Valid Parenthesis
//  */
//
function isValid(s: string): boolean {
  let stack: string[] = [];

  for (const p of s) {
    const last = stack.at(-1);
    if (last === "(" && p === ")") {
      stack.pop();
    } else if (last === "[" && p === "]") {
      stack.pop();
    } else if (last === "{" && p === "}") {
      stack.pop();
    } else {
      stack.push(p);
    }
  }
  return stack.length === 0;
}

let str = "()";
console.log("Example 1:", isValid(str)); // true

str = "()[]{}";
console.log("Example 2:", isValid(str)); // true

str = "(]";
console.log("Example 3:", isValid(str)); // false

File: cpp/leetcode/20.cpp
// 20. Valid Parentheses
#include <string>
#include <stack>
#include <unordered_map>
#include <cassert>

using namespace std;

class Solution {
    public:
    bool isValid(string s) {
        stack<char> st = {};
        unordered_map<char, char> m = {{ ')', '(' }, { '}', '{' }, { ']', '[' }};

        for (char c: s) {
            if (m.find(c) == m.end()) {
                st.push(c);
            } else if (!st.empty() && st.top() == m[c]) {
                st.pop();
            } else {
                return false;
            }
        }

        return st.empty();
    }
};

int main() {
    Solution s;
    string s1 = "()[]{}";
    bool expected1 = true;
    auto result1 = s.isValid(s1);
    assert(result1 == expected1);
    return 0;
}
